File: hibernate/src/main/java/io/github/robertomike/hefesto/builders/Hefesto.java
package io.github.robertomike.hefesto.builders;

import io.github.robertomike.hefesto.actions.Join;
import io.github.robertomike.hefesto.actions.JoinFetch;
import io.github.robertomike.hefesto.actions.wheres.WhereCustom;
import io.github.robertomike.hefesto.actions.wheres.WhereField;
import io.github.robertomike.hefesto.enums.JoinOperator;
import io.github.robertomike.hefesto.enums.Operator;
import io.github.robertomike.hefesto.enums.WhereOperator;
import io.github.robertomike.hefesto.constructors.*;
import io.github.robertomike.hefesto.models.HibernateModel;
import io.github.robertomike.hefesto.models.BaseModel;
import io.github.robertomike.hefesto.utils.Page;
import jakarta.persistence.criteria.*;
import org.hibernate.QueryException;
import org.hibernate.Session;
import org.hibernate.query.Query;

import java.util.*;

public class Hefesto<T extends BaseModel> extends BaseBuilder<T, Session, ConstructWhereImplementation, ConstructJoinImplementation<T>, ConstructOrderImplementation, ConstructSelectImplementation<T>, Hefesto<T>> {
    private final ConstructJoinFetch joinsFetch = new ConstructJoinFetch();
    private Class<?> originalModel = null;
    private Class<?> customResultSubQuery = null;

    public Hefesto(Class<T> model) {
        super(model);
        orders = new ConstructOrderImplementation();
        wheres = new ConstructWhereImplementation();
        joins = new ConstructJoinImplementation<>();
        selects = new ConstructSelectImplementation<>();
    }

    public Hefesto(Class<? extends HibernateModel> model, Class<T> resultClass) {
        super(resultClass);
        this.originalModel = model;
        orders = new ConstructOrderImplementation();
        wheres = new ConstructWhereImplementation();
        joins = new ConstructJoinImplementation<>();
        selects = new ConstructSelectImplementation<>();
    }

    /**
     * Create a new instance of the Hefesto class with the given model.
     *
     * @param  model  the model class to be used
     * @return        a new instance of Hefesto with the given model
     */
    public static <T extends BaseModel> Hefesto<T> make(Class<T> model) {
        return new Hefesto<>(model);
    }


    /**
     * Adds a join to load in the query.
     *
     * @param  field    the field to join on
     * @param  alias    this alias is not for the query is only when you need to use the join in a where condition or select
     * @return          the updated Hefesto object
     */
    public Hefesto<T> join(String field, String alias) {
        joins.add(Join.make(field, alias));
        return this;
    }

    /**
     * Adds a join to load in the query.
     *
     * @param  field    the field to join on
     * @param  operator the join operator
     * @return          the updated Hefesto object
     */
    public Hefesto<T> join(String field, JoinOperator operator) {
        joins.add(Join.make(field, operator));
        return this;
    }

    /**
     * Adds a join to load in the query.
     *
     * @param  field  the field to join on
     * @return        the updated Hefesto instance
     */
    public Hefesto<T> join(String field) {
        joins.add(Join.make(field));
        return this;
    }

    /**
     * This method adds a where field to the query that allows you to filter by a field of a join, the same root or the parent when is a subQuery
     * @param parentField could be parent field or a join field
     */
    public Hefesto<T> whereField(String field, String parentField) {
        getWheres().add(new WhereField(field, parentField));
        return this;
    }

    /**
     * This method adds a where that allow you to pass lambda and return a Predicate
     */
    public Hefesto<T> whereCustom(WhereCustom.Custom custom) {
        getWheres().add(new WhereCustom(custom));
        return this;
    }

    /**
     * This method adds a where with or that allow you to pass lambda and return a Predicate
     */
    public Hefesto<T> orWhereCustom(WhereCustom.Custom custom) {
        getWheres().add(new WhereCustom(custom, WhereOperator.OR));
        return this;
    }

    /**
     * This method adds a where field to the query that allows you to filter by a field of a join, the same root or the parent when is a subQuery
     * @param operator not all operators are supported, only LIKE, NOT_LIKE, EQUAL, DIFF, GREATER, LESS, GREATER_OR_EQUAL, LESS_OR_EQUAL
     * @param parentField could be parent field or a join field
     */
    public Hefesto<T> whereField(String field, Operator operator, String parentField) {
        getWheres().add(new WhereField(field, operator, parentField));
        return this;
    }

    /**
     * This method will allow you to preload the relationships you want to fetch
     */
    public Hefesto<T> with(JoinFetch... relationship) {
        joinsFetch.addAll(relationship);
        return this;
    }

    /**
     * This method will allow you to preload the relationships you want to fetch
     */
    public Hefesto<T> with(String... relationships) {
        for (var relationship : relationships) {
            joinsFetch.add(JoinFetch.make(relationship));
        }
        return this;
    }

    /**
     * This method will allow you to preload the relationships you want to fetch with the join type
     */
    public Hefesto<T> with(String relationship, JoinType joinType) {
        joinsFetch.add(JoinFetch.make(relationship, joinType));
        return this;
    }

    /**
     * Returns an Optional containing the first element of the result set,
     * or an empty Optional if the result set is empty.
     *
     * @return          an Optional containing the first element of the result set,
     *                  or an empty Optional if the result set is empty.
     */
    @Override
    public Optional<T> findFirst() {
        this.limit = 1;
        return Optional.ofNullable(createQuery().getSingleResultOrNull());
    }

    /**
     * Creates a query for the given criteria.
     *
     * @return          The created query.
     */
    protected Query<T> createQuery() {
        var cb = getSession().getCriteriaBuilder();
        var cr = cb.createQuery(model);
        var root = getRoot(cr);

        joinsFetch.construct(root);

        joins.construct(root);
        selects.setJoins(joins.getJoins());
        if (originalModel == null) {
            selects.construct((Root<T>) root, cr, cb);
        } else {
            selects.multiSelect(root, cr, cb);
        }
        wheres.setJoins(joins.getJoins()).construct(cb, cr, root);
        orders.setJoins(joins.getJoins()).construct(cb, cr, root);

        var query = getSession().createQuery(cr);

        if (limit != null) {
            query.setMaxResults(limit);
        }
        if (offset != null) {
            query.setFirstResult(offset);
        }

        return query;
    }

    /**
     * Retrieves the root entity for the given CriteriaQuery.
     *
     * @param  cr  the CriteriaQuery object to retrieve the root from
     * @return     the root entity for the given CriteriaQuery
     */
    private Root<?> getRoot(CriteriaQuery<T> cr) {
        if (originalModel == null) {
            return cr.from(model);
        }
        return cr.from(originalModel);
    }

    /**
     * Generates a sub-query for the given criteria query, root, criteria builder, and joins.
     *
     * @param  cr     the criteria query
     * @param  parentRoot  the parent root
     * @param  cb     the criteria builder
     * @param  parentJoins  the joins
     * @return        the generated subquery
     */
    public Subquery<?> getSubQuery(CriteriaQuery<?> cr, Root<?> parentRoot, CriteriaBuilder cb, Map<String, jakarta.persistence.criteria.Join<?, ?>> parentJoins) {
        var sub = customResultSubQuery != null ? cr.subquery(customResultSubQuery) : cr.subquery(model);
        var root = sub.from(model);

        joins.construct(root);
        var allJoins = new HashMap<>(parentJoins);
        allJoins.putAll(joins.getJoins());
        wheres.setJoins(allJoins).constructSubQuery(sub, cb, root, parentRoot);
        selects.setJoins(allJoins).constructSubQuery(root, sub);

        return sub;
    }

    public Hefesto<T> setCustomResultForSubQuery(Class<?> customResultSubQuery) {
        this.customResultSubQuery = customResultSubQuery;
        return this;
    }

    public boolean hasCustomResultForSubQuery() {
        return customResultSubQuery != null;
    }

    /**
     * Retrieves a list of objects.
     *
     * @return  a list of objects
     */
    public List<T> get() {
        return createQuery().getResultList();
    }

    /**
     * Retrieves a page of results from the database based on the specified limit and offset.
     *
     * @param  limit   the maximum number of results to retrieve
     * @param  offset  the starting position of the results
     * @return         a Page object containing the retrieved results, the offset used, and the total number of results
     */
    public Page<T> page(int limit, long offset) {
        this.offset = Math.toIntExact(offset);
        this.limit = limit;

        var total = countResults();

        return new Page<>(createQuery().getResultList(), offset, total);
    }

    /**
     * Counts the number of results based on the given criteria.
     *
     * @return the count of results as a Long value
     */
    public Long countResults() {
        var cb = getSession().getCriteriaBuilder();
        var cr = cb.createQuery(Long.class);
        var root = cr.from(model);

        cr.select(cb.count(root));
        joins.construct(root);
        wheres.setJoins(joins.getJoins()).construct(cb, cr, root);

        return getSession().createQuery(cr).getSingleResult();
    }

    /**
     * Find the first result of the specified result class.
     *
     * @param  resultClass  the class of the result to be returned
     * @return              the first result of the specified class, or null if no result is found
     */
    public <R> R findFirstFor(Class<R> resultClass) {
        if (selects.isEmpty()) {
            throw new QueryException("You need put at least one select");
        }

        var cr = commonConstructForCustomResult(resultClass);

        var query = getSession().createQuery(cr);
        query.setMaxResults(1);

        return query.getSingleResultOrNull();
    }

    /**
     * Finds and retrieves a list of objects of the specified resultClass.
     *
     * @param  resultClass   the class of the objects to be retrieved
     * @return               a list of objects of the specified resultClass
     */
    public <R> List<R> findFor(Class<R> resultClass) {
        if (selects.isEmpty()) {
            throw new QueryException("You need put at least one select");
        }

        var cr = commonConstructForCustomResult(resultClass);

        return getSession().createQuery(cr).getResultList();
    }

    /**
     * Generates a common criteria query for custom result.
     *
     * @param  resultClass  the class of the result
     * @return              the generated criteria query
     */
    private <R> CriteriaQuery<R> commonConstructForCustomResult(Class<R> resultClass) {
        var cb = getSession().getCriteriaBuilder();
        var cr = cb.createQuery(resultClass);
        var root = cr.from(model);

        joins.construct(root);
        selects.setJoins(joins.getJoins()).multiSelect(root, cr, cb);
        wheres.setJoins(joins.getJoins()).construct(cb, cr, root);
        orders.setJoins(joins.getJoins()).construct(cb, cr, root);

        return cr;
    }
}

File: hibernate-query-language/src/main/kotlin/io/github/robertomike/hefesto/hql/builders/Hefesto.kt
package io.github.robertomike.hefesto.hql.builders

import io.github.robertomike.hefesto.hql.actions.wheres.WhereRaw
import io.github.robertomike.hefesto.builders.BaseBuilder
import io.github.robertomike.hefesto.enums.JoinOperator
import io.github.robertomike.hefesto.enums.WhereOperator
import io.github.robertomike.hefesto.hql.constructors.ConstructGroupByImplementation
import io.github.robertomike.hefesto.hql.constructors.ConstructJoinFetch
import io.github.robertomike.hefesto.hql.constructors.ConstructJoinImplementation
import io.github.robertomike.hefesto.hql.constructors.ConstructOrderImplementation
import io.github.robertomike.hefesto.hql.constructors.ConstructSelectImplementation
import io.github.robertomike.hefesto.hql.constructors.ConstructWhereImplementation
import io.github.robertomike.hefesto.models.BaseModel
import io.github.robertomike.hefesto.utils.Page
import io.github.robertomike.hefesto.utils.SharedMethods
import io.github.robertomike.hefesto.hql.utils.SubQueryContext
import org.hibernate.Session
import org.hibernate.query.Query
import java.util.*
import java.util.function.Consumer

class Hefesto<T : BaseModel> : BaseBuilder<T, Session, ConstructWhereImplementation, ConstructJoinImplementation,
        ConstructOrderImplementation, ConstructSelectImplementation, ConstructGroupByImplementation, Hefesto<T>>,
        SharedMethods<Hefesto<T>> {

    override val joinsFetch = ConstructJoinFetch()
    var acronymTable: String = ""
    private var originalModel: Class<out BaseModel>? = null

    // Lazy initialization of executor - created only when query execution is needed
    private val executor: HefestoExecutor<T> by lazy {
        HefestoExecutor(model, table!!, acronymTable)
    }

    constructor(model: Class<T>) : super(model) {
        setup()
    }

    constructor(modelEntity: Class<out BaseModel>, dto: Class<T>) : super(dto) {
        originalModel = modelEntity
        setup()
    }

    private fun setup() {
        orders = ConstructOrderImplementation()
        wheres = ConstructWhereImplementation()
        joins = ConstructJoinImplementation()
        selects = ConstructSelectImplementation()
        groupBy = ConstructGroupByImplementation()
        table = (originalModel ?: model).simpleName
        acronymTable = table!!.lowercase()
    }

    /**
     * Create a new instance of the Hefesto class with the given model.
     *
     * @param model the model class to be used
     * @return a new instance of Hefesto with the given model
     */
    companion object {
        @JvmStatic
        fun <T : BaseModel> make(model: Class<T>): Hefesto<T> {
            return Hefesto(model)
        }
    }

    /**
     * Unsupported method - use join with Join object instead
     *
     * @param table     the table to join
     * @param joinField the field to join on
     * @param operator  the join operator
     * @return return the current instance
     */
    @Suppress("UNUSED_PARAMETER")
    fun joinUnsupported(table: String, joinField: String, operator: JoinOperator): Hefesto<T> {
        throw UnsupportedOperationException("This methods are not supported")
    }

    /**
     * This method is not supported in HQL version
     */
    override fun join(table: String, joinField: String, operator: JoinOperator): Hefesto<T> {
        throw UnsupportedOperationException("This method is not supported in HQL version")
    }

    /**
     * This method adds a where that allow you to pass lambda and return a Predicate
     */
    fun whereRaw(custom: WhereRaw): Hefesto<T> {
        wheres.add(custom)
        return this
    }

    /**
     * This method adds a where that allow you to pass lambda and return a Predicate
     */
    fun whereRaw(raw: String): Hefesto<T> {
        wheres.add(WhereRaw(raw))
        return this
    }

    /**
     * This method adds a where with or that allow you to pass lambda and return a Predicate
     */
    fun orWhereRaw(raw: String): Hefesto<T> {
        wheres.add(WhereRaw(raw, WhereOperator.OR))
        return this
    }

    /**
     * Returns an Optional containing the first element of the result set,
     * or an empty Optional if the result set is empty.
     *
     * @return an Optional containing the first element of the result set,
     * or an empty Optional if the result set is empty.
     */
    override fun findFirst(): Optional<T> {
        return executor.findFirst(getSessionInstance(), selects, wheres, joins, joinsFetch, orders, groupBy, this)
    }

    /**
     * Creates a query for the given criteria.
     *
     * @return The created query.
     */
    fun <R> createQuery(): Query<R> {
        return executor.createQuery(getSessionInstance(), selects, wheres, joins, joinsFetch, orders, groupBy, limit, offset, this)
    }

    fun <R> createBaseQuery(): Query<R> {
        return executor.createBaseQuery(getSessionInstance(), selects, wheres, joins, joinsFetch, orders, groupBy, limit, offset, false, this)
    }

    /**
     * Retrieves a list of objects.
     *
     * @return a list of objects
     */
    override fun get(): List<T> {
        return executor.get(getSessionInstance(), selects, wheres, joins, joinsFetch, orders, groupBy, limit, offset, this)
    }

    /**
     * Retrieves a page of results from the database based on the specified limit and offset.
     *
     * @param limit  the maximum number of results to retrieve
     * @param offset the starting position of the results
     * @return a Page object containing the retrieved results, the offset used, and the total number of results
     */
    override fun page(limit: Int, offset: Long): Page<T> {
        return executor.page(getSessionInstance(), selects, wheres, joins, joinsFetch, orders, groupBy, limit, offset, this)
    }

    /**
     * Counts the number of results based on the given criteria.
     *
     * @return the count of results as a Long value
     */
    override fun countResults(): Long {
        return executor.countResults(getSessionInstance(), wheres, joins, groupBy, this)
    }

    /**
     * Find the first result of the specified result class.
     *
     * @param resultClass the class of the result to be returned
     * @return the first result of the specified class, or null if no result is found
     */
    fun <R> findFirstFor(resultClass: Class<R>): R {
        return executor.findFirstFor(getSessionInstance(), resultClass, selects, wheres, joins, joinsFetch, orders, groupBy, this)
    }

    /**
     * Finds and retrieves a list of objects of the specified resultClass.
     *
     * @param resultClass the class of the objects to be retrieved
     * @return a list of objects of the specified resultClass
     */
    fun <R> findFor(resultClass: Class<R>): List<R> {
        return executor.findFor(getSessionInstance(), resultClass, selects, wheres, joins, joinsFetch, orders, groupBy, limit, offset, this)
    }

    fun getQuery(params: MutableMap<String, Any?>): String {
        return executor.getQuery(selects, wheres, joins, joinsFetch, orders, groupBy, params, false, this)
    }

    fun getSubQuery(params: MutableMap<String, Any?>): String {
        return executor.getSubQuery(selects, wheres, joins, orders, groupBy, params, this)
    }

    // ================================
    // Lambda-based Subquery Methods
    // ================================

    /**
     * Adds a WHERE IN clause using a lambda-based subquery.
     *
     * Usage:
     * ```java
     * var users = Hefesto.make(User.class)
     *     .whereIn("id", UserPet.class, subQuery -> {
     *         subQuery.addSelect("user.id");
     *         subQuery.where("petType", "DOG");
     *     })
     *     .get();
     * ```
     *
     * @param field The field to check against the subquery results
     * @param subQueryModel The entity class for the subquery
     * @param configurer Lambda to configure the subquery
     * @return this builder for chaining
     */
    fun <S : BaseModel> whereIn(field: String, subQueryModel: Class<S>, configurer: Consumer<SubQueryContext<S>>): Hefesto<T> {
        val subQuery = make(subQueryModel)
        val context = SubQueryContext(subQuery)
        configurer.accept(context)
        return whereIn(field, subQuery)
    }

    /**
     * Adds a WHERE NOT IN clause using a lambda-based subquery.
     *
     * Usage:
     * ```java
     * var users = Hefesto.make(User.class)
     *     .whereNotIn("id", UserPet.class, subQuery -> {
     *         subQuery.addSelect("user.id");
     *         subQuery.where("petType", "CAT");
     *     })
     *     .get();
     * ```
     *
     * @param field The field to check against the subquery results
     * @param subQueryModel The entity class for the subquery
     * @param configurer Lambda to configure the subquery
     * @return this builder for chaining
     */
    fun <S : BaseModel> whereNotIn(field: String, subQueryModel: Class<S>, configurer: Consumer<SubQueryContext<S>>): Hefesto<T> {
        val subQuery = make(subQueryModel)
        val context = SubQueryContext(subQuery)
        configurer.accept(context)
        return whereNotIn(field, subQuery)
    }

    // Note: whereExists/whereNotExists are not supported in HQL base builder
    // Use whereRaw() for custom EXISTS queries if needed

    /**
     * Adds an OR WHERE IN clause using a lambda-based subquery.
     *
     * @param field The field to check against the subquery results
     * @param subQueryModel The entity class for the subquery
     * @param configurer Lambda to configure the subquery
     * @return this builder for chaining
     */
    fun <S : BaseModel> orWhereIn(field: String, subQueryModel: Class<S>, configurer: Consumer<SubQueryContext<S>>): Hefesto<T> {
        val subQuery = make(subQueryModel)
        val context = SubQueryContext(subQuery)
        configurer.accept(context)
        return orWhereIn(field, subQuery)
    }

    /**
     * Adds an OR WHERE NOT IN clause using a lambda-based subquery.
     *
     * @param field The field to check against the subquery results
     * @param subQueryModel The entity class for the subquery
     * @param configurer Lambda to configure the subquery
     * @return this builder for chaining
     */
    fun <S : BaseModel> orWhereNotIn(field: String, subQueryModel: Class<S>, configurer: Consumer<SubQueryContext<S>>): Hefesto<T> {
        val subQuery = make(subQueryModel)
        val context = SubQueryContext(subQuery)
        configurer.accept(context)
        return orWhereNotIn(field, subQuery)
    }

    // Note: orWhereExists/orWhereNotExists are not supported in HQL base builder
    // Use whereRaw() or orWhereRaw() for custom EXISTS queries if needed

}
